### PR TITLE
feat: Update target objects for privilege-granting resources

### PR DESCRIPTION
**Pull request by sfc-gh-jcieslak**
A follow-up for https://github.com/Snowflake-Labs/terraform-provider-snowflake/pull/2604#discussion_r1519554878. Updated target objects for privilege-granting resources based on documentation (also added references above available object's list).

Pr also contains:
- #2689 is fixed in the pr and acceptance tests are added. 
- #2696 documentation fix

### CODE REVIEW

File: pkg/sdk/object_types.go [hunk 1]
```diff
@@ -68,6 +68,10 @@ const (
 	ObjectTypeHybridTable          ObjectType = "HYBRID TABLE"
 	ObjectTypeImageRepository      ObjectType = "IMAGE REPOSITORY"
 	ObjectTypeProjectionPolicy     ObjectType = "PROJECTION POLICY"
+	ObjectTypeDataMetricFunction   ObjectType = "DATA METRIC FUNCTION"
```
Comment (sfc-gh-asawicki) [inline pkg/sdk/object_types.go:71]: Have you checked that they does not have to be added to any other resources? Please check it.

**Review by sfc-gh-asawicki — CHANGES_REQUESTED** [review]

File: docs/resources/tag_association.md [hunk 2]
```diff
@@ -81,7 +81,7 @@ resource "snowflake_tag_association" "column_association" {
 ### Required
 
 - `object_identifier` (Block List, Min: 1) Specifies the object identifier for the tag association. (see [below for nested schema](#nestedblock--object_identifier))
-- `object_type` (String) Specifies the type of object to add a tag to. ex: 'ACCOUNT', 'COLUMN', 'DATABASE', etc. For more information: https://docs.snowflake.com/en/user-guide/object-tagging.html#supported-objects
+- `object_type` (String) Specifies the type of object to add a tag. Allowed object types: [                              ].
```
Comment (sfc-gh-asawicki) [inline docs/resources/tag_association.md:84]: empty?

File: pkg/resources/tag_association.go [hunk 3]
```diff
@@ -18,6 +19,68 @@ import (
 	snowflakeValidation "github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/validation"
 )
 
+var (
```
Comment (sfc-gh-asawicki) [inline pkg/resources/tag_association.go:22]: why here and not in SDK? (like ValidGrantOwnershipObjectTypesString and the rest?)

**Review by sfc-gh-asawicki — CHANGES_REQUESTED** [review]

File: pkg/resources/tag_association.go (hunk 3 above)
Comment (sfc-gh-jcieslak) [reply pkg/resources/tag_association.go:22]: Moved to a new file in sdk tag_association_validations (similarly to what we have in grants)

File: docs/resources/tag_association.md (hunk 2 above)
Comment (sfc-gh-jcieslak) [reply docs/resources/tag_association.md:84]: Not sure why it was this way, but after moving the validations it generated correctly (most likely it was something with `init` function and moving it to the other package that created some kind of dependency).

**Review by sfc-gh-asawicki — APPROVED** [review]

**Pull request merged**